### ISSUE TITLE
Each except variable should have an independent type and scope

### ISSUE THREAD

**Issue by JukkaL** [JukkaL/mypy#1045]
Mypy complains about this code:

```
try:
    pass
except IOError as e:
    pass
except RuntimeError as e:
    pass
```

It doesn't like the second except block:

```
t.py:5: error: Incompatible types in assignment (expression has type "RuntimeError", variable has type "IOError")
```

However, the variable `e` is in scope after the try statement, which is wrong.

The workaround is to rename the second exception variable, but mypy should allow the code as such.  Each except block should define a new local variable.

**Comment by gvanrossum** [comment]
Right, although technically it's not a new scope -- the variable is just
deleted, in Python 3 at least. In Python 2 it actually survives. :-(

On Thursday, December 3, 2015, Jukka Lehtosalo <notifications@github.com>
wrote:

> Mypy complains about this code:
>
> try:
>     pass
> except IOError as e:
>     pass
> except RuntimeError as e:
>     pass
>
> It doesn't like the second except block:
>
> t.py:5: error: Incompatible types in assignment (expression has type "RuntimeError", variable has type "IOError")
>
> However, the variable e is in scope after the try statement, which is
> wrong.
>
> The workaround is to rename the second exception variable, but mypy should
> allow the code as such. Each except block should define a new local
> variable.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JukkaL/mypy/issues/1045>.
>


-- 
--Guido (mobile)

**Comment by JukkaL** [comment]
Ah hmm. I guess mypy could enforce that the exception variable isn't used outside the except block even in Python 2, though this would give some bogus errors.

As the variable is just deleted, then we may want to introduce a new 'bottom' or undefined type for variables that is generated as the result of `del x` (and after an except block). Reading a variable with this type would be considered an error, but it would valid as an lvalue. 

This would also let us support cases like this:

```
x = 1
del x
x  # mypy should complain that 'x' may be undefined here
```

**Comment by gvanrossum** [comment]
Yeah, all good ideas. I like enforcing that it isn't used in PY2 outside
the except clause (it's a regular cause for confusion).

**Comment by JukkaL** [comment]
Fixed by @ecprice (#1054).
